### PR TITLE
Generalize fan following DMap

### PR DESCRIPTION
**Pull request by treeowl**
Generalize `fan` to `fang` (yeah, that's a horrible name) allowing fanning `DMap`s with arbitrary targets (not just `Identity`). I haven't yet dug down enough to work out how to do the same for `merge`, but it would be useful.

### CODE REVIEW

**Comment by ryantrinkle** [comment]
Awesome!  This all looks great; maybe we could rename it to `fanG` and merge?

**Comment by treeowl** [comment]
That's fine by me. I don't really like that name either, but ... whatever.
Do you have any suggestions for tackling that for merge? It looks a lot
hairier, and I don't even know where to start.

**Comment by treeowl** [comment]
Renamed. Should be ready to merge.

File: src/Reflex/Class.hs [hunk 1]
```diff
@@ -298,6 +302,18 @@ class ( MonadHold t (PushM t)
   mergeIntIncremental :: Incremental t (PatchIntMap (Event t a)) -> Event t (IntMap a)
   fanInt :: Event t (IntMap a) -> EventSelectorInt t a
 
+-- | Efficiently fan-out an event to many destinations. You should save the
+-- result in a @let@-binding, and then repeatedly 'select' on the result to
+-- create child events
+fan :: forall t k. (Reflex t, GCompare k)
+    => Event t (DMap k Identity) -> EventSelector t k
+    --TODO: Can we help enforce the partial application discipline here?  The combinator is worthless without it
+fan e = EventSelector (fixup (selectg (fanG e) :: k a -> Event t (Identity a)) :: forall a. k a -> Event t a)
```
Comment (treeowl) [inline src/Reflex/Class.hs]: Should we match on the `Coercion` and then produce a lambda, or something like that, to make sure there's only one `eventCoercion` call per `Reflex` instance?

Comment (treeowl) [reply src/Reflex/Class.hs]: It probably doesn't actually matter...

Comment (Ericson2314) [reply src/Reflex/Class.hs]: I was thinking about such things and opened https://github.com/reflex-frp/reflex/issues/319

**Comment by cgibbard** [comment]
If we're going with fanG, probably best to have it be selectG for
consistency.

Also, it's worth thinking about whether it wouldn't be a more reasonable
option to just make this generalisation just be fan itself -- the number of
occurrences of the fully general 'fan' in the wild is pretty small, and
this is a straightforward generalisation.

**Comment by ryantrinkle** [comment]
The deprecation plan shouldn't burn our more hardcore users, but it might
make sense to eventually come around to fan=fanG

**Comment by Ericson2314** [comment]
We'll have the same thing again with NonEmptyDMap for fan and merge, so I say new name for now, and replace `fan` later.

**Comment by ali-abrar** [comment]
Could you please update the changelog?

**Comment by treeowl** [comment]
@ali-abrar, updated.

**Comment by saulzar** [comment]
Generalize fan to fanG allowing fanning DMaps with functors other than Identity.

**Pull request merged**